lib/spack/llnl/util/cpu.py

```diff
@@ -0,0 +1,316 @@
+            features (list of str): supported CPU flags. Note that the semantic
+                of the flags in this field might vary among architectures, if
+                at all present. For instance x86_64 processors will list all
+                the flags supported by a given CPU while Arm processors will
+                list instead only the flags that have been added on top of the
+                base model for the current micro-architecture.
+            compilers (dict): compiler support to generate tuned code for this
+                micro-architecture.
+            generation (int): generation of the micro-architecture, if
+                relevant.
+        """
+        self.name = name
+        self.parents = parents
+        self.ancestors = parents[:]
+        for parent in parents:
+            self.ancestors.extend(
+                list(filter(lambda a: a not in self.ancestors,
+                            parent.ancestors))
+            )
+        self.vendor = vendor
+        self.features = features
+        self.compilers = compilers
+        self.generation = generation
+
+    def _ensure_strictly_orderable(self, other):
+        if not (self in other.ancestors or other in self.ancestors):
+            msg = "There is no ordering relationship between targets "
+            msg += "%s and %s." % (self.name, other.name)
+            raise ValueError(msg)
+
+    def __eq__(self, other):
+        if not isinstance(other, MicroArchitecture):
+            return NotImplemented
+
+        return (self.name == other.name and
+                self.vendor == other.vendor and
+                self.features == other.features and
+                self.ancestors == other.ancestors and
+                self.compilers == other.compilers and
+                self.generation == other.generation)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __lt__(self, other):
+        if not isinstance(other, MicroArchitecture):
+            return NotImplemented
+
+        self._ensure_strictly_orderable(other)
+
+        # If the current micro-architecture is in the list of ancestors
+        # of the other micro-architecture, then it's less than the other
+        if self in other.ancestors:
+            return True
+
+        return False
+
+    def __le__(self, other):
+        return (self == other) or (self < other)
+
+    def __gt__(self, other):
+        return not (self <= other)
+
+    def __ge__(self, other):
+        return not (self < other)
+
+    def __repr__(self):
+        cls_name = self.__class__.__name__
+        fmt = cls_name + '({0.name!r}, {0.parents!r}, {0.vendor!r}, ' \
+                         '{0.features!r}, {0.compilers!r}, {0.generation!r})'
+        return fmt.format(self)
+
+    def __str__(self):
+        return self.name
+
+    def __contains__(self, feature):
+        # The semantic here looks for the exact feature name
+        return feature in self.features
```

Comment:
> To address both this and the ancestors issue, would it not make sense to propagate all features forward to children in the graph, instead of having children store all their ancestors?

Are we ensured that there won't be cases where an instruction is removed from one micro-architecture to the next one? If we forward features it might become harder or more complex to remove a feature if we need it.

Reply:
> Question for @becker33 and @alalazo: it will be useful for users to be able to say, e.g, if 'sse3' in target, but with this semantic, that will fail if the model supports ssse3.

At the moment `cpu.MicroArchitecture` has a strict semantic on features and `__contains__` returns true checking only the features used for detection. The behavior you ask for is currently implemented in `spack.architecture.Target`. The rationale is that:

 - `cpu.Microarchitecture` is a lower level abstraction on which `spack.architecture.Target` is built
- Packagers will only interact with `Target` so in the end the semantic you ask for will still be respected in packages

Do I understand it correctly that you still want this behavior pushed down to `cpu.Microarchitecture`?

Reply:
> Are we ensured that there won't be cases where an instruction is removed from one micro-architecture to the next one? If we forward features it might become harder or more complex to remove a feature if we need it.

Yeah this is a good point.  I think the aliases handle this case just fine without our needed to push features down, and they let us deal with special cases like this.  I think any successful API for this will need to avoid assuming any uniformity on the part of the vendors 😬.

> Do I understand it correctly that you still want this behavior pushed down to `cpu.Microarchitecture`?

Yes.  I don't think this means we need to make the original features unrecoverable (like you could query them or have a property for them or something), but I do think that microarchitecture should be providing most of the semantics.  Having the aliases in `Target` makes me worry that we're pulling in a lot of arch detail into the `Spec`, which isn't where it should go.

Personally, I think the `archspec` library we pull out of this should act a lot like Spack specs, but only handle the very limited case of comparing architectures for compatibility.  Instead of satisfies we just have `compatible()` or something.  It's stuff everyone will need to implement on top anyway, and it makes the library way more useful.

Reply:
I pushed the logic down to `cpu.MicroArchitecture` in 28b2d96935581b8bfb7b8e3c1c3d39d81412dcbe